source/solvers/vof.cc

```diff
@@ -1082,5 +1116,219 @@ VolumeOfFluid<dim>::assemble_mass_matrix_diagonal(
+                                       update_values |
+                                         update_quadrature_points);
+
+  const unsigned int n_q_points = this->cell_quadrature->size();
+
+  const FEValuesExtractors::Scalar pressure(dim);
+  std::vector<double>              pressure_values(n_q_points);
+  std::vector<double>              phase_values(n_q_points);
+
+  unsigned int boundary_id =
+    this->simulation_parameters.boundary_conditions.id[i_bc];
+
+  // Useful definitions for readability
+  const unsigned int density_fluid_0 =
+    this->simulation_parameters.physical_properties.fluids[0].density;
+  const unsigned int density_fluid_1 =
+    this->simulation_parameters.physical_properties.fluids[1].density;
+  const unsigned int wetting_threshold =
+    this->simulation_parameters.boundary_conditions_vof.wetting_threshold[i_bc];
+  const unsigned int peeling_threshold =
+    this->simulation_parameters.boundary_conditions_vof.peeling_threshold[i_bc];
+
+  // Loop on cell_vof
+  for (const auto &cell_vof : dof_handler.active_cell_iterators())
+    {
+      if (cell_vof->is_locally_owned() && cell_vof->at_boundary())
+        {
+          for (const auto face : cell_vof->face_indices())
+            {
+              if (cell_vof->face(face)->at_boundary() &&
+                  cell_vof->face(face)->boundary_id() == boundary_id)
+                {
+                  // Equivalent dof_handler for CFD
+                  typename DoFHandler<dim>::active_cell_iterator cell_cfd(
+                    &(*(this->triangulation)),
+                    cell_vof->level(),
+                    cell_vof->index(),
+                    dof_handler_cfd);
+
+                  // Reinit fe_face_values for CFD and VOF
+                  fe_face_values_vof.reinit(cell_vof, face);
+                  fe_face_values_cfd.reinit(cell_cfd, face);
+
+                  // Get pressure values
+                  fe_face_values_cfd[pressure].get_function_values(
+                    current_solution_cfd, pressure_values);
+
+                  // Get phase values
+                  fe_face_values_vof.get_function_values(present_solution,
+                                                         phase_values);
+
+                  // Number of points meeting the peeling criteria in the cell
+                  unsigned int nb_q_peeled(0);
+
+                  // Number of points meeting the wetting criteria in the cell
+                  unsigned int nb_q_wet(0);
+
+                  // Local index (see deal.II step 4)
+                  cell_vof->get_dof_indices(dof_indices_vof);
+
+                  for (unsigned int q = 0; q < n_q_points; q++)
+                    {
+                      if (pressure_values[q] > wetting_threshold)
+                        {
+                          // wetting of lowest density fluid
+                          if ((density_fluid_1 > density_fluid_0) &&
+                              (phase_values[q] < 0.5))
```

Comment:
It would be less confusing if the less dense and the dense fluids where pre-established and these ifs in the quadrature point logic were removed.

Reply:
I am not sure that I understand you: do you suggest that "fluid 1" would always be set to be the densest one? In this case we would have to ask the user to follow this when writting the prm file, it did not seem appropriate

Reply:
no I meant like... instead of having all of these if in the loop, you could define a variable called "critical phase value" or something and pre-define it.

Reply:
I did that at first, but it occurred to me that it did not change much the readability, because I would trade this if condition `if ((density_fluid_1 > density_fluid_0) && (phase_values[q] < 0.5))` for `if ((wetting_phase_value == 1) && (phase_values[q] < 0.5))`, which is not a lot clearer. Besides, I would have an if loop prior to that to define the `wetting_phase_value` and `peeling_phase_value`, so it seemed not a good trade. Maybe we could talk briefly about it to clarify that?